### DISCUSSION TITLE
Longpress and rotate encoder

### DISCUSSION THREAD

**Discussion by dsg25** [tttapa/Control-Surface#753] [Q&A]
Hello!
Help, please.
Why don't work?

```
#include <Control_Surface.h> // Include the Control Surface library
#include <Arduino_Helpers.h> // https://github.com/tttapa/Arduino-Helpers
 
#include <AH/Hardware/MultiPurposeButton.hpp>

int CCNote = 0x10;

// Instantiate a MIDI over USB interface.
HairlessMIDI_Interface midi;

MultiPurposeButton btn{2};

CCRotaryEncoder enc1 {
 {5, 6},       // pins
 CCNote, // MIDI address (CC number + optional channel)
 1,            // optional multiplier if the control isn't fast enough
 };

void setup() {
  btn.setLongPressDelay(1000);
  btn.setMultiPressDelay(400);
  btn.begin();
  Control_Surface.begin(); // Initialize Control Surface (calls button.begin())
  
}

void loop() {
  buttonEnc();
  Control_Surface.loop(); // Update the Control Surface (calls button.update())
}

void buttonEnc() {
  auto longPressTime = btn.getLongPressedTime();
  switch (btn.update()) {
    case btn.None: break;
    case btn.PressStart: 
      break;
    case btn.LongPress: CCNote = 0x11;
    case btn.LongPressRelease: CCNote = 0x10;
  
      break;
    case btn.MultiPress:
    
      break;
    case btn.MultiPressDone:
  
      break;
  }

}
```

**Comment by tttapa** [comment]
> ```cpp
> CCRotaryEncoder enc1 {
>   {5, 6},       // pins
>   CCNote, // MIDI address (CC number + optional channel)
>   1,            // optional multiplier if the control isn't fast enough
> };
> ```

Here, a copy of the value of `CCNote` is made: `enc1` and `CCNote` do not get linked in any way, and changing the value of `CCNote` afterwards does not (and cannot) affect `enc1`.

To change the MIDI address at runtime, consider using a Bank: https://tttapa.github.io/Control-Surface-doc/Doxygen/da/dc1/FAQ.html#faq-change-address-runtime

You also have implicit fallthrough in your switch, you need more `break`s.

**Reply by dsg25** [reply]
> > 
> To change the MIDI address at runtime, consider using a Bank: https://tttapa.github.io/Control-Surface-doc/Doxygen/da/dc1/FAQ.html#faq-change-address-runtime
> 

Sorry, I don't understand how to make it work:

```
Bankable::CCRotaryEncoder enc {
  {bank, BankType::CHANGE_CHANNEL},     // bank configuration
  2,3                                   //  pin
  {MIDI_CC::Channel_Volume, CHANNEL_1}, // address
};
```

**Reply by tttapa** [reply]
The encoder pins should be passed as a single argument, so add braces around `2,3`, see https://tttapa.github.io/Control-Surface-doc/Doxygen/da/d40/classBankable_1_1CCRotaryEncoder.html#a9ca816053c6c26b789c2b9ac62fafaea.

```cpp
Bankable::CCRotaryEncoder enc {
  {bank, BankType::CHANGE_CHANNEL},     // bank configuration
  {2, 3}                                // pins
  {MIDI_CC::Channel_Volume, CHANNEL_1}, // address
};
```

**Reply by dsg25** [reply]
Compilation Error:

```
#include <Control_Surface.h> // Include the Control Surface library

// Instantiate a MIDI over USB interface
USBMIDI_Interface midi;

// Instantiate four Banks, with two tracks per bank. 
// Compare these numbers to the diagram above.
Bank<4> bank(1);
//   │       └───── number of tracks per bank
//   └───────────── number of banks

// Instantiate a Bank selector to control which one of the four Banks is active.
IncrementDecrementSelector<4> selector {
    bank,       // Bank to manage
    {11, 12},     // push button pins (increment, decrement)
    Wrap::Wrap, // Wrap around
};


Bankable::CCRotaryEncoder enc {
  {bank, BankType::CHANGE_CHANNEL},     // bank configuration
  {2, 3}                                // pins
  {MIDI_CC::Channel_Volume, CHANNEL_1}, // address
};


void setup() {
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
}
```



> 
> Bank:24:3: error: expected '}' before '{' token
>    {MIDI_CC::Channel_Volume, CHANNEL_1}, // address
>    ^
> Bank:24:3: error: no matching function for call to 'CS::Bankable::CCRotaryEncoder::CCRotaryEncoder(<brace-enclosed initializer list>)'
> In file included from /Users/kocha/Documents/Arduino/libraries/Control-Surface-master/src/Control_Surface.h:84:0,
>                  from /var/folders/hl/q60hqqvd2ts18dx83zyd5ydr0000gn/T/arduino_modified_sketch_65857/Bank.ino:2:
> /Users/kocha/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Bankable/CCRotaryEncoder.hpp:54:5: note: candidate: CS::Bankable::CCRotaryEncoder::CCRotaryEncoder(CS::OutputBankConfig<>, CS::Encoder&&, CS::MIDIAddress, int16_t, uint8_t)
>      CCRotaryEncoder(OutputBankConfig<> config, Encoder &&encoder,
>      ^~~~~~~~~~~~~~~
> /Users/kocha/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Bankable/CCRotaryEncoder.hpp:54:5: note:   candidate expects 5 arguments, 2 provided
> /Users/kocha/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Bankable/CCRotaryEncoder.hpp:20:7: note: candidate: CS::Bankable::CCRotaryEncoder::CCRotaryEncoder(CS::Bankable::CCRotaryEncoder&&)
>  class CCRotaryEncoder
>        ^~~~~~~~~~~~~~~
> /Users/kocha/Documents/Arduino/libraries/Control-Surface-master/src/MIDI_Outputs/Bankable/CCRotaryEncoder.hpp:20:7: note:   candidate expects 1 argument, 2 provided
> Bank:24:3: error: expected ',' or ';' before '{' token
>    {MIDI_CC::Channel_Volume, CHANNEL_1}, // address
>    ^
> Bank:24:39: error: expected unqualified-id before ',' token
>    {MIDI_CC::Channel_Volume, CHANNEL_1}, // address
>                                        ^
> Bank:25:1: error: expected unqualified-id before '}' token
>  };
>  ^
> Bank:25:1: error: expected declaration before '}' token
> exit status 1
> expected '}' before '{' token

**Reply by tttapa** [reply]
My bad, I lost a comma during copy -pasting. Add a comma after `{2, 3}`, as suggested by the compiler.

**Reply by dsg25** [reply]
Hi!
I dont understand, how in this code use one LED for selection bank.
Also, Dont  find example use IncrementSelectorLEDs.

```
#include <Control_Surface.h> // Include the Control Surface library

int CCNote = 0x10;

// Instantiate a MIDI over USB interface.
USBMIDI_Interface midi;



Bank<2> bank(1);
//   │       └───── number of tracks per bank
//   └───────────── number of banks


IncrementSelector<2> selector {
    bank,       // Bank to manage
    2,     // push button pins (increment, decrement)
     // Wrap around
};

Bankable::CCRotaryEncoder enc3 {
  {bank, BankType::CHANGE_ADDRESS},     // bank configuration
  {5, 6},                                // pins
  {CCNote}, // address
};


void setup() {
 
  Control_Surface.begin(); // Initialize Control Surface (calls button.begin())
  
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface (calls button.update())
}

```
 Thank!

**Reply by tttapa** [reply]
You can simply query [`bank.getSelection()`](https://tttapa.github.io/Control-Surface-doc/Doxygen/db/dbd/classBank.html#a18abc6b96728c56b1b4260c507afda7a) in the loop and use that to turn an LED on/off.

You can compare the documentation for the constructors of IncrementSelector and IncrementSelectorLEDs: the latter has an additional argument for the LEDs. Arguably, IncrementSelectorLEDs is not that useful, because it always uses two LEDs.
